Add _WKWebExtensionContext and WebKit::WebExtensionContext.

xeenon opened the pull request:
#### 476ad6dbea8e2e3d0fe1f72e4c9de6bfcc73189a
<pre>
Add _WKWebExtensionContext and WebKit::WebExtensionContext.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246255">https://bugs.webkit.org/show_bug.cgi?id=246255</a>

Reviewed by Brian Weinstein.

This class represents the runtime environment for an extension, which tracks permission access,
and will eventually handle things like background page loading, injecting content, storage access, etc.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toAPI):
(WebKit::toImpl):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/WebExtensionContextIdentifier.h: Added.
* Source/WebKit/Shared/WebExtensionContextParameters.h: Added.
* Source/WebKit/Shared/WebExtensionContextParameters.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension requestedPermissions]):
(-[_WKWebExtension optionalPermissions]):
(-[_WKWebExtension requestedPermissionMatchPatterns]):
(-[_WKWebExtension optionalPermissionMatchPatterns]):
(-[_WKWebExtension allRequestedMatchPatterns]):
(-[_WKWebExtension requestedPermissionOrigins]): Deleted.
(-[_WKWebExtension optionalPermissionOrigins]): Deleted.
(-[_WKWebExtension allRequestedOrigins]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm: Added.
(+[_WKWebExtensionContext contextWithExtension:]):
(-[_WKWebExtensionContext initWithExtension:]):
(-[_WKWebExtensionContext dealloc]):
(-[_WKWebExtensionContext extension]):
(-[_WKWebExtensionContext extensionController]):
(-[_WKWebExtensionContext baseURL]):
(-[_WKWebExtensionContext setBaseURL:]):
(-[_WKWebExtensionContext uniqueIdentifier]):
(-[_WKWebExtensionContext setUniqueIdentifier:]):
(toImpl):
(-[_WKWebExtensionContext grantedPermissions]):
(-[_WKWebExtensionContext setGrantedPermissions:]):
(-[_WKWebExtensionContext grantedPermissionMatchPatterns]):
(-[_WKWebExtensionContext setGrantedPermissionMatchPatterns:]):
(-[_WKWebExtensionContext deniedPermissions]):
(-[_WKWebExtensionContext setDeniedPermissions:]):
(-[_WKWebExtensionContext deniedPermissionMatchPatterns]):
(-[_WKWebExtensionContext setDeniedPermissionMatchPatterns:]):
(-[_WKWebExtensionContext requestedOptionalAccessToAllHosts]):
(-[_WKWebExtensionContext setRequestedOptionalAccessToAllHosts:]):
(toAPI):
(-[_WKWebExtensionContext currentPermissions]):
(-[_WKWebExtensionContext currentPermissionMatchPatterns]):
(-[_WKWebExtensionContext hasPermission:]):
(-[_WKWebExtensionContext hasPermission:inTab:]):
(-[_WKWebExtensionContext hasAccessToURL:]):
(-[_WKWebExtensionContext hasAccessToURL:inTab:]):
(-[_WKWebExtensionContext permissionStateForPermission:]):
(-[_WKWebExtensionContext permissionStateForPermission:inTab:]):
(-[_WKWebExtensionContext setPermissionState:forPermission:]):
(-[_WKWebExtensionContext setPermissionState:forPermission:expirationDate:]):
(-[_WKWebExtensionContext permissionStateForURL:]):
(-[_WKWebExtensionContext permissionStateForURL:inTab:]):
(-[_WKWebExtensionContext setPermissionState:forURL:]):
(-[_WKWebExtensionContext setPermissionState:forURL:expirationDate:]):
(-[_WKWebExtensionContext permissionStateForMatchPattern:]):
(-[_WKWebExtensionContext permissionStateForMatchPattern:inTab:]):
(-[_WKWebExtensionContext setPermissionState:forMatchPattern:]):
(-[_WKWebExtensionContext setPermissionState:forMatchPattern:expirationDate:]):
(-[_WKWebExtensionContext hasAccessToAllURLs]):
(-[_WKWebExtensionContext hasAccessToAllHosts]):
(-[_WKWebExtensionContext userGesturePerformedInTab:]):
(-[_WKWebExtensionContext hasActiveUserGestureInTab:]):
(-[_WKWebExtensionContext cancelUserGestureForTab:]):
(-[_WKWebExtensionContext _apiObject]):
(-[_WKWebExtensionContext _webExtensionContext]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContextInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContextPrivate.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTab.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasRequestedPermission const):
(WebKit::WebExtension::supportedPermissions):
(WebKit::WebExtension::requestedPermissions):
(WebKit::WebExtension::optionalPermissions):
(WebKit::WebExtension::requestedPermissionMatchPatterns):
(WebKit::WebExtension::optionalPermissionMatchPatterns):
(WebKit::WebExtension::allRequestedMatchPatterns):
(WebKit::WebExtension::populatePermissionsPropertiesIfNeeded):
(WebKit::toAPI):
(WebKit::WebExtension::requestedPermissionOrigins): Deleted.
(WebKit::WebExtension::optionalPermissionOrigins): Deleted.
(WebKit::WebExtension::allRequestedOrigins): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm: Added.
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::setBaseURL):
(WebKit::WebExtensionContext::isURLForThisExtension):
(WebKit::WebExtensionContext::setUniqueIdentifier):
(WebKit::WebExtensionContext::grantedPermissions):
(WebKit::WebExtensionContext::setGrantedPermissions):
(WebKit::WebExtensionContext::deniedPermissions):
(WebKit::WebExtensionContext::setDeniedPermissions):
(WebKit::WebExtensionContext::grantedPermissionMatchPatterns):
(WebKit::WebExtensionContext::setGrantedPermissionMatchPatterns):
(WebKit::WebExtensionContext::deniedPermissionMatchPatterns):
(WebKit::WebExtensionContext::setDeniedPermissionMatchPatterns):
(WebKit::WebExtensionContext::postAsyncNotification):
(WebKit::WebExtensionContext::grantPermissions):
(WebKit::WebExtensionContext::denyPermissions):
(WebKit::WebExtensionContext::grantPermissionMatchPatterns):
(WebKit::WebExtensionContext::denyPermissionMatchPatterns):
(WebKit::WebExtensionContext::removeGrantedPermissions):
(WebKit::WebExtensionContext::removeGrantedPermissionMatchPatterns):
(WebKit::WebExtensionContext::removeDeniedPermissions):
(WebKit::WebExtensionContext::removeDeniedPermissionMatchPatterns):
(WebKit::WebExtensionContext::removePermissions):
(WebKit::WebExtensionContext::removePermissionMatchPatterns):
(WebKit::WebExtensionContext::removeExpired):
(WebKit::WebExtensionContext::hasPermission):
(WebKit::WebExtensionContext::permissionState):
(WebKit::WebExtensionContext::setPermissionState):
(WebKit::WebExtensionContext::clearCachedPermissionStates):
(WebKit::WebExtensionContext::hasAccessToAllURLs):
(WebKit::WebExtensionContext::hasAccessToAllHosts):
(WebKit::WebExtensionContext::userGesturePerformed):
(WebKit::WebExtensionContext::hasActiveUserGesture const):
(WebKit::WebExtensionContext::cancelUserGesture):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::validSchemes):
(WebKit::WebExtensionMatchPattern::supportedSchemes):
(WebKit::patternCache):
(WebKit::WebExtensionMatchPattern::isSupported const):
(WebKit::WebExtensionMatchPattern::expandedStrings const):
(WebKit::WebExtensionMatchPattern::isValidScheme):
(WebKit::WebExtensionMatchPattern::matchesURL):
(WebKit::WebExtensionMatchPattern::matchesPattern):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::wrapper const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp: Added.
(WebKit::WebExtensionContext::get):
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::~WebExtensionContext):
(WebKit::WebExtensionContext::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Added.
(WebKit::WebExtensionContext::create):
(WebKit::WebExtensionContext::identifier const):
(WebKit::WebExtensionContext::isLoaded const):
(WebKit::WebExtensionContext::extension const):
(WebKit::WebExtensionContext::extensionController const):
(WebKit::WebExtensionContext::baseURL const):
(WebKit::WebExtensionContext::uniqueIdentifier const):
(WebKit::WebExtensionContext::requestedOptionalAccessToAllHosts const):
(WebKit::WebExtensionContext::setRequestedOptionalAccessToAllHosts):
(WebKit::WebExtensionContext::currentPermissions):
(WebKit::WebExtensionContext::currentPermissionMatchPatterns):
(WebKit::WebExtensionContext::hasPermission):
(WebKit::WebExtensionContext::permissionState):
(WebKit::WebExtensionContext::wrapper const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in: Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
(WebKit::WebExtensionMatchPattern::wrapper const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp: Added.
(WebKit::WebExtensionContextProxy::getOrCreate):
(WebKit::WebExtensionContextProxy::WebExtensionContextProxy):
(WebKit::WebExtensionContextProxy::~WebExtensionContextProxy):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h: Added.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in: Added.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.h: Include header to fix build.
* Source/WebKit/WebProcess/WebCoreSupport/WebScreenOrientationManager.cpp: Ditto.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionContext.mm: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/255374@main">https://commits.webkit.org/255374@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46cb7a25b501037186f60a062bf578cb933d40d5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92333 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1563 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22919 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/102105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1559 "Built successfully") | [❌ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29946 "Failed to checkout and rebase branch from PR 5177") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97994 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/84760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/36364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/34127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/17741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/38001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1684 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/39901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/36884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -0,0 +1,153 @@
+#import <Foundation/Foundation.h>
+
+#import <WebKit/_WKWebExtensionPermission.h>
+#import <WebKit/_WKWebExtensionMatchPattern.h>
```
Nit: Sorting.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
```diff
@@ -0,0 +1,650 @@
+
+@implementation _WKWebExtensionContext
+
+#if ENABLE(WK_WEB_EXTENSIONS)
```
Nit: Should all of the constants be in the `ENABLE` block? Or do they not need to be?

xeenon replied:
They need to be exported symbols, since the header is not behind a guard since guards can't be in the public headers.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
```diff
@@ -0,0 +1,650 @@
+    NSMutableDictionary<_WKWebExtensionPermission, NSDate *> *result = [NSMutableDictionary dictionaryWithCapacity:permissions.size()];
+
+    for (auto& entry : permissions)
+        [result setObject:WebKit::toAPI(entry.value) forKey:entry.key];
```
Nit: Could this use subscript notation?

xeenon replied:
Yes! But, it requires casting the WTF String to `NSString *`. The `setObject:forKey:` version auto casts and does the conversion, so it is cleaner IMO.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
```diff
@@ -0,0 +1,650 @@
+    NSMutableDictionary<_WKWebExtensionMatchPattern *, NSDate *> *result = [NSMutableDictionary dictionaryWithCapacity:permissionMatchPatterns.size()];
+
+    for (auto& entry : permissionMatchPatterns)
+        [result setObject:WebKit::toAPI(entry.value) forKey:wrapper(entry.key)];
```
Ditto about subscript notation?

xeenon replied:
Yep! But see above.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
```diff
@@ -0,0 +1,650 @@
+}
+
+#else // ENABLE(WK_WEB_EXTENSIONS)
+
```
Is this a stub class used for testing?

b-weinstein replied:
(The implementation of the methods below when the `WK_WEB_EXTENSIONS` enable flag is off)

xeenon replied:
Sort of. They need implemented for platforms that don't have the flag enabled, like Watch and TV.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -0,0 +1,825 @@
+    auto urlMatchesWildcardHostPatterns = ^(WebExtensionMatchPattern& pattern) {
+        if (!pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesURL(url);
```
Is there a time where `pattern.matchesURL` would be false if the pattern matches all hosts?

xeenon replied:
If the pattern is `"http://*/*"` and the scheme or path don't match still.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -0,0 +1,825 @@
+    auto urlMatchesWildcardHostPatterns = ^(WebExtensionMatchPattern& pattern) {
+        if (!pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesPattern(pattern);
```
Ditto my previous question about this check being necessary.

Also - `pattern.matchesPattern(pattern)` is concerning - which value will pattern have here?

xeenon replied:
Good catch! I think I need to rename the arg!

(Similar to the URL one, the scheme or path might not match.)

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -0,0 +1,825 @@
+    auto urlMatchesPatternIgnoringWildcardHostPatterns = ^(WebExtensionMatchPattern& pattern) {
+        if (pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesPattern(pattern);
```
What will happen with `pattern` as the block argument vs `pattern` as the function argument here?

xeenon replied:
I'll rename the arg.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
```diff
@@ -45,25 +45,23 @@
 {
-    // FIXME: Add extension schemes to this set.
-    static NSSet<NSString *> *schemes = [NSSet setWithObjects:@"*", @"http", @"https", @"file", @"ftp", nil];
+    // FIXME: Add custom extension schemes to this set somehow.
```
Super nit: Not sure the `somehow` adds too much here.

xeenon replied:
It is kind of a hard problem to figure out still. 😅

b-weinstein commented on Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.h:
```diff
@@ -27,6 +27,7 @@
 
 #if PLATFORM(COCOA) && ENABLE(GPU_PROCESS) && ENABLE(MEDIA_STREAM)
 
+#include "GPUProcessConnection.h"
```
What happened here?

xeenon replied:
Build fix since adding files to Source.txt sometimes uncovers missing header imports.

b-weinstein commented on Source/WebKit/WebProcess/WebCoreSupport/WebScreenOrientationManager.cpp:
```diff
@@ -26,6 +26,7 @@
 #include "config.h"
 #include "WebScreenOrientationManager.h"
 
+#include "WebPage.h"
```
And here?

xeenon replied:
Same. Build fix.

xeenon committed:
Add _WKWebExtensionContext and WebKit::WebExtensionContext.
https://bugs.webkit.org/show_bug.cgi?id=246255

Reviewed by Brian Weinstein.

This class represents the runtime environment for an extension, which tracks permission access,
and will eventually handle things like background page loading, injecting content, storage access, etc.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toAPI):
(WebKit::toImpl):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/WebExtensionContextIdentifier.h: Added.
* Source/WebKit/Shared/WebExtensionContextParameters.h: Added.
* Source/WebKit/Shared/WebExtensionContextParameters.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension requestedPermissions]):
(-[_WKWebExtension optionalPermissions]):
(-[_WKWebExtension requestedPermissionMatchPatterns]):
(-[_WKWebExtension optionalPermissionMatchPatterns]):
(-[_WKWebExtension allRequestedMatchPatterns]):
(-[_WKWebExtension requestedPermissionOrigins]): Deleted.
(-[_WKWebExtension optionalPermissionOrigins]): Deleted.
(-[_WKWebExtension allRequestedOrigins]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm: Added.
(+[_WKWebExtensionContext contextWithExtension:]):
(-[_WKWebExtensionContext initWithExtension:]):
(-[_WKWebExtensionContext dealloc]):
(-[_WKWebExtensionContext extension]):
(-[_WKWebExtensionContext extensionController]):
(-[_WKWebExtensionContext baseURL]):
(-[_WKWebExtensionContext setBaseURL:]):
(-[_WKWebExtensionContext uniqueIdentifier]):
(-[_WKWebExtensionContext setUniqueIdentifier:]):
(toImpl):
(-[_WKWebExtensionContext grantedPermissions]):
(-[_WKWebExtensionContext setGrantedPermissions:]):
(-[_WKWebExtensionContext grantedPermissionMatchPatterns]):
(-[_WKWebExtensionContext setGrantedPermissionMatchPatterns:]):
(-[_WKWebExtensionContext deniedPermissions]):
(-[_WKWebExtensionContext setDeniedPermissions:]):
(-[_WKWebExtensionContext deniedPermissionMatchPatterns]):
(-[_WKWebExtensionContext setDeniedPermissionMatchPatterns:]):
(-[_WKWebExtensionContext requestedOptionalAccessToAllHosts]):
(-[_WKWebExtensionContext setRequestedOptionalAccessToAllHosts:]):
(toAPI):
(-[_WKWebExtensionContext currentPermissions]):
(-[_WKWebExtensionContext currentPermissionMatchPatterns]):
(-[_WKWebExtensionContext hasPermission:]):
(-[_WKWebExtensionContext hasPermission:inTab:]):
(-[_WKWebExtensionContext hasAccessToURL:]):
(-[_WKWebExtensionContext hasAccessToURL:inTab:]):
(-[_WKWebExtensionContext permissionStateForPermission:]):
(-[_WKWebExtensionContext permissionStateForPermission:inTab:]):
(-[_WKWebExtensionContext setPermissionState:forPermission:]):
(-[_WKWebExtensionContext setPermissionState:forPermission:expirationDate:]):
(-[_WKWebExtensionContext permissionStateForURL:]):
(-[_WKWebExtensionContext permissionStateForURL:inTab:]):
(-[_WKWebExtensionContext setPermissionState:forURL:]):
(-[_WKWebExtensionContext setPermissionState:forURL:expirationDate:]):
(-[_WKWebExtensionContext permissionStateForMatchPattern:]):
(-[_WKWebExtensionContext permissionStateForMatchPattern:inTab:]):
(-[_WKWebExtensionContext setPermissionState:forMatchPattern:]):
(-[_WKWebExtensionContext setPermissionState:forMatchPattern:expirationDate:]):
(-[_WKWebExtensionContext hasAccessToAllURLs]):
(-[_WKWebExtensionContext hasAccessToAllHosts]):
(-[_WKWebExtensionContext userGesturePerformedInTab:]):
(-[_WKWebExtensionContext hasActiveUserGestureInTab:]):
(-[_WKWebExtensionContext cancelUserGestureForTab:]):
(-[_WKWebExtensionContext _apiObject]):
(-[_WKWebExtensionContext _webExtensionContext]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContextInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContextPrivate.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTab.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasRequestedPermission const):
(WebKit::WebExtension::supportedPermissions):
(WebKit::WebExtension::requestedPermissions):
(WebKit::WebExtension::optionalPermissions):
(WebKit::WebExtension::requestedPermissionMatchPatterns):
(WebKit::WebExtension::optionalPermissionMatchPatterns):
(WebKit::WebExtension::allRequestedMatchPatterns):
(WebKit::WebExtension::populatePermissionsPropertiesIfNeeded):
(WebKit::toAPI):
(WebKit::WebExtension::requestedPermissionOrigins): Deleted.
(WebKit::WebExtension::optionalPermissionOrigins): Deleted.
(WebKit::WebExtension::allRequestedOrigins): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm: Added.
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::setBaseURL):
(WebKit::WebExtensionContext::isURLForThisExtension):
(WebKit::WebExtensionContext::setUniqueIdentifier):
(WebKit::WebExtensionContext::grantedPermissions):
(WebKit::WebExtensionContext::setGrantedPermissions):
(WebKit::WebExtensionContext::deniedPermissions):
(WebKit::WebExtensionContext::setDeniedPermissions):
(WebKit::WebExtensionContext::grantedPermissionMatchPatterns):
(WebKit::WebExtensionContext::setGrantedPermissionMatchPatterns):
(WebKit::WebExtensionContext::deniedPermissionMatchPatterns):
(WebKit::WebExtensionContext::setDeniedPermissionMatchPatterns):
(WebKit::WebExtensionContext::postAsyncNotification):
(WebKit::WebExtensionContext::grantPermissions):
(WebKit::WebExtensionContext::denyPermissions):
(WebKit::WebExtensionContext::grantPermissionMatchPatterns):
(WebKit::WebExtensionContext::denyPermissionMatchPatterns):
(WebKit::WebExtensionContext::removeGrantedPermissions):
(WebKit::WebExtensionContext::removeGrantedPermissionMatchPatterns):
(WebKit::WebExtensionContext::removeDeniedPermissions):
(WebKit::WebExtensionContext::removeDeniedPermissionMatchPatterns):
(WebKit::WebExtensionContext::removePermissions):
(WebKit::WebExtensionContext::removePermissionMatchPatterns):
(WebKit::WebExtensionContext::removeExpired):
(WebKit::WebExtensionContext::hasPermission):
(WebKit::WebExtensionContext::permissionState):
(WebKit::WebExtensionContext::setPermissionState):
(WebKit::WebExtensionContext::clearCachedPermissionStates):
(WebKit::WebExtensionContext::hasAccessToAllURLs):
(WebKit::WebExtensionContext::hasAccessToAllHosts):
(WebKit::WebExtensionContext::userGesturePerformed):
(WebKit::WebExtensionContext::hasActiveUserGesture const):
(WebKit::WebExtensionContext::cancelUserGesture):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::validSchemes):
(WebKit::WebExtensionMatchPattern::supportedSchemes):
(WebKit::patternCache):
(WebKit::WebExtensionMatchPattern::isSupported const):
(WebKit::WebExtensionMatchPattern::expandedStrings const):
(WebKit::WebExtensionMatchPattern::isValidScheme):
(WebKit::WebExtensionMatchPattern::matchesURL):
(WebKit::WebExtensionMatchPattern::matchesPattern):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::wrapper const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp: Added.
(WebKit::WebExtensionContext::get):
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::~WebExtensionContext):
(WebKit::WebExtensionContext::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Added.
(WebKit::WebExtensionContext::create):
(WebKit::WebExtensionContext::identifier const):
(WebKit::WebExtensionContext::isLoaded const):
(WebKit::WebExtensionContext::extension const):
(WebKit::WebExtensionContext::extensionController const):
(WebKit::WebExtensionContext::baseURL const):
(WebKit::WebExtensionContext::uniqueIdentifier const):
(WebKit::WebExtensionContext::requestedOptionalAccessToAllHosts const):
(WebKit::WebExtensionContext::setRequestedOptionalAccessToAllHosts):
(WebKit::WebExtensionContext::currentPermissions):
(WebKit::WebExtensionContext::currentPermissionMatchPatterns):
(WebKit::WebExtensionContext::hasPermission):
(WebKit::WebExtensionContext::permissionState):
(WebKit::WebExtensionContext::wrapper const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in: Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
(WebKit::WebExtensionMatchPattern::wrapper const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp: Added.
(WebKit::WebExtensionContextProxy::getOrCreate):
(WebKit::WebExtensionContextProxy::WebExtensionContextProxy):
(WebKit::WebExtensionContextProxy::~WebExtensionContextProxy):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h: Added.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in: Added.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.h: Include header to fix build.
* Source/WebKit/WebProcess/WebCoreSupport/WebScreenOrientationManager.cpp: Ditto.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionContext.mm: Added.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/255374@main
